protocol.go

```diff
@@ -284,13 +284,22 @@ func isCommaOrSpace(c rune) bool {
 }
 
 func discard(reader io.Reader) (int64, error) {
-	if lr, ok := reader.(*io.LimitedReader); ok {
-		return io.Copy(io.Discard, lr)
+	lreader, ok := reader.(*io.LimitedReader)
+	if !ok {
+		lreader = &io.LimitedReader{R: reader, N: discardLimit}
 	}
+	limit := lreader.N
 	// We don't want to get stuck throwing data away forever, so limit how much
 	// we're willing to do here.
-	lr := &io.LimitedReader{R: reader, N: discardLimit}
-	return io.Copy(io.Discard, lr)
+	wroteN, err := io.Copy(io.Discard, lreader)
+	if errors.Is(err, io.EOF) {
+		err = nil
+	}
+	if wroteN == limit {
+		// Ensure we error if we hit the limit.
+		err = io.ErrShortBuffer
+	}
```

Comment:
Technically, the limit and the payload could be the same size, in which case this error would be spurious. You'd need to either set the limit reader to `discardLimit+1` or follow up with an attempt to read a single byte to decide on if this error is appropriate.